### ISSUE TITLE
Wrong type detected

### ISSUE THREAD

**Issue by HeinrichAD** [microsoft/pyright#3659]
Issue Type: <b>Bug</b>

# Behaviour
## Expected vs. Actual

The existence of an `__init__` should not change the detected type.

_Expected_: If `__new__` is overwritten and returns a `str` it should be detected as `str`, even if `__init__` is overwritten.
_Actual_: As soon as `__init__` is also overwritten, it will be not be detected as `str` but as the original class.

The consequences are the usual ones if the type is recognized incorrectly. AutoComplete displays the wrong elements and if one relies on it, it comes to runtime errors.

## Steps to reproduce:

```python
#!/usr/bin/env python3

class A:
    def __new__(cls) -> str:
        return "Hello World!"

class B:
    def __new__(cls) -> str:
        return "Hello World!"
    def __init__(self):
        pass

a = A()  # detected as (variable) a: str
b = B()  # detected as (variable) b: B    # <---- this should be detected as str as well

print(type(a))  # <class 'str'>
print(type(b))  # <class 'str'>
```

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.4
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Venv
-   Value of the `python.languageServer` setting: Default

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
[Info  - 3:23:17 PM] (41708) Pylance language server 2022.6.30 (pyright 212d1465) starting
[Info  - 3:23:17 PM] (41708) Server root directory: /home/user/.vscode-server/extensions/ms-python.vscode-pylance-2022.6.30/dist
Notebook support: Legacy
[Info  - 3:23:17 PM] (41708) No configuration file found.
[Info  - 3:23:17 PM] (41708) No pyproject.toml file found.
[Info  - 3:23:17 PM] (41708) Setting pythonPath for service "Test": "/home/user/projects/Test/.venv/bin/python"
[Warn  - 3:23:17 PM] (41708) stubPath /home/user/projects/Test/typings is not a valid directory.
[Info  - 3:23:17 PM] (41708) Assuming Python version 3.7
[Info  - 3:23:17 PM] (41708) Assuming Python platform Linux
[Info  - 3:23:17 PM] (41708) Searching for source files
[Info  - 3:23:17 PM] (41708) Found 31 source files
[Info  - 3:23:17 PM] (41708) Background analysis(1) root directory: /home/user/.vscode-server/extensions/ms-python.vscode-pylance-2022.6.30/dist
[Info  - 3:23:17 PM] (41708) Background analysis(1) started
```

</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

languageServer: "Pylance"

```

</p>
</details>

Extension version: 2022.8.1
VS Code version: Code 1.68.1 (30d9c6cd9483b2cc586687151bcbcd635f373630, 2022-06-15T02:58:26.441Z)
OS version: Linux x64 5.18.5-arch1-1
Restricted Mode: No
Remote OS version: Linux x64 4.15.0-151-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz (8 x 2301)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 1, 2|
|Memory (System)|15.34GB (0.55GB free)|
|Process Argv|--unity-launch --crash-reporter-id ab2207cb-01ee-4aca-bbf1-3ce6e3a080c6|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|i3|
|XDG_CURRENT_DESKTOP|i3|
|XDG_SESSION_DESKTOP|i3|
|XDG_SESSION_TYPE|x11|

|Item|Value|
|---|---|
|Remote|SSH: 139.6.237.137|
|OS|Linux x64 4.15.0-151-generic|
|CPUs|Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz (56 x 1202)|
|Memory (System)|251.80GB (71.45GB free)|
|VM|0%|
</details>

**Comment by erictraut** [comment]
What you're doing here is highly unusual. Can you explain why you are defining a new class but returning an instance of a different class from `__new__`? While this is allowed in Python, it is not something I would recommend, and it's not something that we fully support in pyright, the static type checker upon which Pylance is built. As you've noticed, we do support this in the case where there's a "bare" `__new__` without a corresponding `__init__` method, but that's more of a side effect of the current implementation, not an intentional design decision.

Static analysis of constructor calls is complicated by many factors including the need to handle `__new__`, `__init__`, and metclass `__call__` methods. All of these can be overloaded, and the signatures can differ (e.g. `__init__` may spell out the various parameters while `__new__` might use `*args, **kwargs`). These complexities place constraints on what static analyzers can do and the order in which they need to check these various methods.

I'll note that mypy also doesn't handle `__new__` methods that return an instance of another class.

So I think it's unlikely that we'd make a change to support this case. Doing so will likely break other commonly-used coding patterns.

**Comment by HeinrichAD** [comment]
This isn't an important bug because as you said it's highly unusual. You can close this issue if you want.

I noticed this issue as I played with the adapter pattern inside python and wanted to know how far I could get.

You could say I wanted to create a class which wraps an single object. Each time this class is accessed but the requested attribute is missing, it should return this attribute from the wrapped object. The code worked, only autocomplete obviously didn't know that the class also "offers" all attributes from the object. For this case I thought I could just overwrite `__new__` with the default behavior but a different return type (a union from the new class and the class from the wrapped object).

Most likely there is a better way to "extend" an object.

---

Pseudo Code:

```python
# package 1
class A:
    def __init__(self, dict):
        ...
    ...
    def a():
        ...
    def b():
        ...

# package 2
class B(A, ...):
    def __init__(self, dict):
        ...
    ...
    def export(self) -> A:
        ...

# no external code anymore

# adapter pattern
class C:
    def __init__(self, a_or_b_obj, name, desc, ...):
        self.obj = a_or_b_obj
        self.name = name
        self.desc = desc
        ...
    def __getattr__(self, attr):
        if super(C, self).hasattr(attr):
            return self[attr]
        else:
            return self.obj[attr]
    def export(self):
        if self.obj.hasattr("export"):
            self.obj = self.obj.export()
        return self

d = {...}
a = A(d)
b = B(d)
c1 = C(a, "A", "desc A")
c2 = C(b, "B", "desc B")

# now I can call my own attributes and the attributes from class A
# but autocomplete obviously does not know it.
print(c1.name)
print(c2.name)
c1.a()  # just like it's the object itself
c2.a()
c1.export()
c2.export()
print(c1.desc)  # still available after export
print(c2.desc)
c1.b()
c2.b()

# The idea to fix autocomplete was to overwrite the definition in __new__.
# __new__ shouldn't do anything different but updating the autocomplete functionality (globally).
```

**Comment by erictraut** [comment]
Note that if you do return a mismatched object for the `__new__` method, the `__init__` method will never be called, nor will any other methods that you have defined in your class. So even if it were legitimate to return a mismatched object, it doesn't make any sense to provide an `__init__` method in this case.

I'm going to close.
